internal/server/deviceflow.go

```diff
@@ -79,29 +79,94 @@ func (a *API) GetDeviceFlowStatus(c *gin.Context, req *api.DeviceFlowStatusReque
 		}, nil
 	}
 
-	if dfar.AccessKey != nil {
+	if dfar.ApprovedWith == 0 {
 		return &api.DeviceFlowStatusResponse{
-			Status:     "confirmed",
+			Status:     "pending",
 			DeviceCode: dfar.DeviceCode,
-			LoginResponse: &api.LoginResponse{
-				UserID:           dfar.AccessKey.IssuedFor,
-				Name:             dfar.AccessKey.IssuedForName,
-				AccessKey:        string(dfar.AccessKeyToken),
-				Expires:          api.Time(dfar.AccessKey.ExpiresAt),
-				OrganizationName: dfar.Organization.Name,
-			},
 		}, nil
 	}
 
+	// TODO (jmorganca): there's a case where a user could log out between approving a request
+	// and the device calling the status endpoint, which would result in the user not
+	// being logged in on the device. We should fix this by querying _all_ access keys
+	// (including deleted ones), but only if the user still exists (in case they were deleted
+	// between approving and this endpoint being called).
+	ak, err := data.GetAccessKey(rctx.DBTxn, data.GetAccessKeysOptions{ByID: dfar.ApprovedWith})
+	if err != nil {
+		return nil, fmt.Errorf("%w: device flow get approve key: %v", internal.ErrUnauthorized, err)
+	}
+
+	// TODO (jmorganca): there's a lot of repeated logic in here and the Login() endpoint
+	// create a new access key for the device
+	accessKey := &models.AccessKey{
+		IssuedFor:     ak.IssuedFor,
+		IssuedForName: ak.IssuedForName,
+
+		// Share the same provider ID that was used to approve
+		ProviderID:          ak.ProviderID,
+		ExpiresAt:           time.Now().UTC().Add(a.server.options.SessionDuration),
+		InactivityTimeout:   time.Now().UTC().Add(a.server.options.SessionInactivityTimeout),
+		InactivityExtension: a.server.options.SessionInactivityTimeout,
+		Scopes:              models.CommaSeparatedStrings{models.ScopeAllowCreateAccessKey},
+	}
+
+	bearer, err := data.CreateAccessKey(rctx.DBTxn, accessKey)
+	if err != nil {
+		return nil, fmt.Errorf("%w: device flow create new key: %v", internal.ErrUnauthorized, err)
+	}
+
+	identity, err := data.GetIdentity(rctx.DBTxn, data.GetIdentityOptions{ByID: ak.IssuedFor})
```

Comment:
nit: might make more sense to do this before creating the access key. The end result is the same in both cases though where we would rollback the transaction on failure.